src/multitenant/mulittenant.service.ts

```diff
@@ -132,16 +134,72 @@ export class MultitenantService {
             Logger.warn(`Attempted to open wallet for walletName ${walletName} from cache but with wrong walletKey`);
         } else {
             Logger.warn(`Attempted to open wallet for walletName ${walletName} but it wasn't in cache`);
+            return await this.handleCacheMiss(walletName, walletKey);
         }
         throw new ProtocolException(ProtocolErrorCode.INTERNAL_SERVER_ERROR, 'Error creating wallet on multitenant');
     }
 
+    /**
+     * If the wallet isn't in the cache we need find the wallet id and token from multitenant aca-py
+     * First we fetch all the wallets and loop through them looking for the one that matches our wallet name
+     * Then we use that wallet id and the passed in wallet key to fetch the token
+     */
+    private async handleCacheMiss(walletName: string, walletKey: string): Promise<Array<string>> {
+        const allWallets = await this.callGetAllWallets();
+        let walletId = null;
+        for (const wallet of allWallets) {
+            if (wallet.settings[`wallet.name`] === walletName) {
+                walletId = wallet.wallet_id;
+                break; // we found our wallet so break out of loop
+            }
+        }
+        if (!walletId) {
+            Logger.warn(`Attempted to open wallet for walletName ${walletName} but it wasn't registered on multitenant`);
+            throw new ProtocolException(ProtocolErrorCode.INTERNAL_SERVER_ERROR, 'Error creating wallet on multitenant');
+        }
+        const token = await this.callGetToken(walletId, walletKey);
+        return [walletId, token];
```

Comment:
Since this token is related to walletid, would it be better to return a tuple instead of an array?

Reply:
Ah yes - good call